### PR TITLE
`@deprecate_arguments` and `@deprecate_function` add deprecation to docstring

### PR DESCRIPTION
**Pull request by Eric-Arellano**
### Summary

Applies https://github.com/Qiskit/qiskit-terra/pull/9685 to our current deprecation decorators.

### CODE REVIEW

File: test/python/utils/test_deprecation.py [hunk 1]
```diff
@@ -15,7 +15,89 @@
...
+            dedent(
+                """\
+
+                .. deprecated:: 9.99
+                  Stop using my_func!
+                """
+            ),
+        )
+
+    def test_deprecate_arguments_runtime_warning(self) -> None:
```
Comment (Eric-Arellano) [inline test/python/utils/test_deprecation.py:74]: None of this functionality is new. But I wanted to make sure I didn't break anything.

File: test/python/utils/test_deprecation.py [hunk 2]
```diff
@@ -15,7 +15,89 @@
...
+            self.assertEqual(new_arg2, "z")
+
+        my_func(new_arg2="z")  # No warnings if no deprecated args used.
+        with self.assertWarnsRegex(DeprecationWarning, "arg1"):
+            my_func(arg1="a", new_arg2="z")
+        with self.assertWarnsRegex(DeprecationWarning, "arg2"):
+            # `arg2` should be converted into `new_arg2`.
+            my_func(arg2="z")  # pylint: disable=missing-kwoa
+
+    def test_deprecate_function_runtime_warning(self) -> None:
```
Comment (Eric-Arellano) [inline test/python/utils/test_deprecation.py:92]: Ditto that this functionality is not new

File: qiskit/algorithms/factorizers/shor.py [hunk 3]
```diff
@@ -57,10 +57,9 @@ class Shor:
     """
 
     @deprecate_function(
-        """The Shor class is deprecated as of Qiskit Terra 0.22.0 and will be removed
-        no sooner than 3 months after the release date.
-        It is replaced by the tutorial at https://qiskit.org/textbook/ch-algorithms/shor.html
-        """,
+        "The Shor class is deprecated as of Qiskit Terra 0.22.0 and will be removed "
```
Comment (1ucian0) [inline qiskit/algorithms/factorizers/shor.py:60]:
```suggestion
        "The Shor class is deprecated and will be removed "
```

Comment (Eric-Arellano) [reply qiskit/algorithms/factorizers/shor.py:60]: This isn't a good change yet -- this is the old `@deprecate_function` decorator which doesn't insert `since` in the message. It doesn't generate the message at all.

A follow up PR will convert this to the new `@deprecate_func`

File: qiskit/algorithms/factorizers/shor.py [hunk 4]
```diff
@@ -57,10 +57,9 @@ class Shor:
     """
 
     @deprecate_function(
-        """The Shor class is deprecated as of Qiskit Terra 0.22.0 and will be removed
```
Comment (Eric-Arellano) [inline qiskit/algorithms/factorizers/shor.py:60]: All this diff does is convert from a multiline string to string concatenation. We can't handle `\n` in `add_deprecation_to_docstring()` (as a simplification)

File: qiskit/utils/deprecation.py [hunk 5]
```diff
@@ -37,15 +37,27 @@ def deprecate_arguments(
...
     """
 
-    del since  # Will be used in a followup to add deprecations to our docs site.
-
     def decorator(func):
+        func_name = func.__qualname__
+        old_kwarg_to_msg = {}
+        for old_arg, new_arg in kwarg_map.items():
+            msg_suffix = (
+                "will in the future be removed." if new_arg is None else f"replaced with {new_arg}."
```
Comment (ElePT) [inline qiskit/utils/deprecation.py:45]: I am not a native speaker, but "will be removed in the future" sounds more natural to me than "will in the future be removed". If there's any specific reason for this order, feel free to ignore my suggestion.

```suggestion
                "will be removed in the future." if new_arg is None else f"replaced with {new_arg}."
```

File: test/python/utils/test_deprecation.py [hunk 6]
```diff
@@ -15,7 +15,89 @@
...
+            dedent(
+                f"""\
+
+                .. deprecated:: 9.99_pending
+                  {my_func.__qualname__} keyword argument old_arg1 is deprecated and replaced with \
+new_arg1.
+
+                .. deprecated:: 9.99_pending
+                  {my_func.__qualname__} keyword argument old_arg2 is deprecated and will in the \
+future be removed.
```
Comment (ElePT) [inline test/python/utils/test_deprecation.py:51]: If you apply my suggestion above, this test will break.

**Review by ElePT — COMMENTED** [review]
I don't have much experience with deprecations, but as a future deprecator the code LGTM! I just have a minor suggestion on the wording of the deprecation message.

File: qiskit/utils/deprecation.py (hunk 5 above)
Comment (Eric-Arellano) [reply qiskit/utils/deprecation.py:45]: Thanks! No particular reason for the old wording. I intentionally didn't change the wording because this helper is going to be deprecated in https://github.com/Qiskit/qiskit-terra/pull/9676, so it doesn't really matter. This PR is mostly just a refactoring.

But there's little downside to improving this wording also.

File: test/python/utils/test_deprecation.py [hunk 7]
```diff
@@ -15,7 +15,89 @@
...
+            {"old_arg1": "new_arg1", "old_arg2": None},
+            category=PendingDeprecationWarning,
+            since="9.99",
+        )
+        def my_func() -> None:
+            pass
+
+        self.assertEqual(
+            my_func.__doc__,
+            dedent(
```
Comment (1ucian0) [inline test/python/utils/test_deprecation.py:42]: not sure about removing idents, as they are kinda important in this context. Do you think that a literal comparison is too strong?

Comment (Eric-Arellano) [reply test/python/utils/test_deprecation.py:42]: This still considers indentation. Note how there are two spaces underneath the line `.. deprecated::`. This is how the docstring renders when there was no prior docstring, which is what we want.

Also, for context, these tests are not focused on ensuring that `add_deprecation_to_docstring` handles all the edge cases of indentation. We already have a whole suite of unit tests for that dedicated purpose. These unit tests here are only to check that we:

1. Generate the right message
2. Actually call the helper function `add_deprecation_to_docstring`

**Review by 1ucian0 — APPROVED** [review]
